passes/source2il.nim

```diff
@@ -42,15 +58,41 @@ template addType(c; kind: NodeKind, body: untyped): uint32 =
     body
   (let r = c.numTypes; inc c.numTypes; r.uint32)
 
+proc parseType(t; n: NodeIndex): TypeKind =
+  case t[n].kind
+  of UnitTy:  tkUnit
+  of BoolTy:  tkBool
+  of IntTy:   tkInt
+  of FloatTy: tkFloat
+  else:       unreachable() # syntax error
+
 proc typeToIL(c; typ: TypeKind): uint32 =
   case typ
+  of tkVoid, tkUnit:
+    unreachable()
   of tkBool:
     c.addType Int: c.types.add(Node(kind: Immediate, val: 1))
   of tkInt, tkError:
     c.addType Int: c.types.add(Node(kind: Immediate, val: 8))
   of tkFloat:
     c.addType Float: c.types.add(Node(kind: Immediate, val: 8))
 
+proc genProcType(c; ret: TypeKind): uint32 =
+  ## Generates a proc type with `ret` as the return type and adds it to `c`.
+  case ret
+  of tkError:
+    unreachable()
+  of tkVoid, tkUnit:
+    # a ``void`` return type means "doesn't return a value", and since there's
```

Comment:
The comment here is aimed at how `unit` is translated. There's no unit type in the `L10`, so a `unit` return type is turned into `(Proc (Void))`, because `Void` there means that the procedure doesn't return a value.

Having said that, removing the `unit` type like this will complicate future work, since it requires all special-casing all `unit` type receivers. I'll change `unit` to be translated to an `(Int 1)` type for now. Later on, an IL with a unit type can be introduced, together with a pass that lowers said IL into one without a `unit` type.

Reply:
> Having said that, removing the `unit` type like this will complicate future work, since it requires all special-casing all `unit` type receivers. I'll change `unit` to be translated to an `(Int 1)` type for now.

I'm guessing you'd rather make that change in a separate PR?

Reply:
It's super simple and not worth the PR overhead -- I'll do it here.